### PR TITLE
feat(TUP-18311): find a way to lower the cost of adding/upgrading BD Distros: have plugin for dynamic config with first implementation to read json / setup jars for jobs.

### PR DESCRIPTION
**Pull request by cmeng-talend**
**Please check if the PR fulfills these requirements**

- [x] The commit(s) message(s) follows our [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md#commit-message-format) ?
- [x] Unit tests for the Java changes have been added (for bug fixes / features) ?
- [ ] TUJ for the JavaJet changes have been added (for bug fixes / features) ?
- [ ] Docs have been added / updated (for bug fixes / features) ?
- [ ] The new code does not introduce new technical issues

**What is the current behavior?** (You can also link to an open issue here)

**What is the new behavior?**

**BREAKING CHANGE**

If this PR contains a breaking change, please describe the impact and migration
path for existing applications.
If not please remove this section.

**Other information**:

### CODE REVIEW

**Comment by cmeng-talend** [comment]
this pull request depends on https://github.com/Talend/tdi-studio-se/pull/1645

File: main/plugins/org.talend.hadoop.distribution.cdh5x/src/org/talend/hadoop/distribution/cdh5x/DynamicCDHDistribution.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+import org.talend.hadoop.distribution.component.ImpalaComponent;
+import org.talend.hadoop.distribution.component.MRComponent;
+import org.talend.hadoop.distribution.component.PigComponent;
+import org.talend.hadoop.distribution.component.SparkBatchComponent;
+import org.talend.hadoop.distribution.component.SparkStreamingComponent;
+import org.talend.hadoop.distribution.component.SqoopComponent;
+import org.talend.hadoop.distribution.dynamic.IDynamicDistribution;
+
+/**
+ * DOC cmeng  class global comment. Detailled comment
```
Comment (amarouni) [inline main/plugins/org.talend.hadoop.distribution.cdh5x/src/org/talend/hadoop/distribution/cdh5x/DynamicCDHDistribution.java]: Could you please add some comments for the class here ?

File: main/plugins/org.talend.hadoop.distribution.cdh5x/src/org/talend/hadoop/distribution/cdh5x/DynamicCDHDistribution.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        };
+        // AbstractDynamicAdapter adapter = DynamicAdapterFactory.getInstance().create(runtimePlugin.getTagName(),
+        // runtimePlugin,
+        // id);
+        // adapter.adapt();
+
+        DynamicServiceUtil.addContribution(bundle, runtimePlugin);
+
+        BundleContext context = bundle.getBundleContext();
+        osgiService = DynamicServiceUtil.registOSGiService(context,
```
Comment (amarouni) [inline main/plugins/org.talend.hadoop.distribution.cdh5x/src/org/talend/hadoop/distribution/cdh5x/DynamicCDHDistribution.java]: New bigdata distribution interfaces need to be added here ?

File: main/plugins/org.talend.hadoop.distribution.cdh5x/src/org/talend/hadoop/distribution/cdh5x/modulegroup/AbstractModuleGroup.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+package org.talend.hadoop.distribution.cdh5x.modulegroup;
+
+import org.apache.commons.lang.StringUtils;
+import org.apache.log4j.Level;
+import org.apache.log4j.LogManager;
+import org.talend.commons.exception.CommonExceptionHandler;
+import org.talend.hadoop.distribution.dynamic.adapter.DynamicPluginAdapter;
+
+/**
+ * DOC cmeng  class global comment. Detailled comment
```
Comment (amarouni) [inline main/plugins/org.talend.hadoop.distribution.cdh5x/src/org/talend/hadoop/distribution/cdh5x/modulegroup/AbstractModuleGroup.java]: Could you please add some comments for the class here ?

File: main/plugins/org.talend.hadoop.distribution/resources/basejars.txt [hunk 4]
```diff
@@ -0,0 +1,233 @@
+HDFS-LIB-DYNAMIC:
```
Comment (amarouni) [inline main/plugins/org.talend.hadoop.distribution/resources/basejars.txt]: Why do we need these base jars here ? And why some of them are `cdh5.10.1`

File: main/plugins/org.talend.hadoop.distribution/src/main/java/org/talend/hadoop/distribution/HadoopDistributionPlugin.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+import java.util.List;
+import java.util.Map;
+
+import org.eclipse.core.runtime.Plugin;
+import org.osgi.framework.BundleContext;
+import org.talend.commons.exception.ExceptionHandler;
+import org.talend.core.model.general.ModuleNeeded;
+import org.talend.librariesmanager.model.ModulesNeededProvider;
+
+
```
Comment (amarouni) [inline main/plugins/org.talend.hadoop.distribution/src/main/java/org/talend/hadoop/distribution/HadoopDistributionPlugin.java]: Some class comments here also can be useful.

File: main/plugins/org.talend.hadoop.distribution.cdh5x/src/org/talend/hadoop/distribution/cdh5x/DynamicCDHDistribution.java (hunk 1 above)
Comment (cmeng-talend) [reply main/plugins/org.talend.hadoop.distribution.cdh5x/src/org/talend/hadoop/distribution/cdh5x/DynamicCDHDistribution.java]: ok, I'll add it

File: main/plugins/org.talend.hadoop.distribution.cdh5x/src/org/talend/hadoop/distribution/cdh5x/DynamicCDHDistribution.java (hunk 2 above)
Comment (cmeng-talend) [reply main/plugins/org.talend.hadoop.distribution.cdh5x/src/org/talend/hadoop/distribution/cdh5x/DynamicCDHDistribution.java]: no, in fact all supported interfaces will be declared in the cdh template class;
I think this class can be used to load/unload all dynamic cdh versions.

File: main/plugins/org.talend.hadoop.distribution.cdh5x/src/org/talend/hadoop/distribution/cdh5x/modulegroup/AbstractModuleGroup.java (hunk 3 above)
Comment (cmeng-talend) [reply main/plugins/org.talend.hadoop.distribution.cdh5x/src/org/talend/hadoop/distribution/cdh5x/modulegroup/AbstractModuleGroup.java]: ok, I'll add it

File: main/plugins/org.talend.hadoop.distribution/resources/basejars.txt (hunk 4 above)
Comment (cmeng-talend) [reply main/plugins/org.talend.hadoop.distribution/resources/basejars.txt]: sorry, this file is no use now, I'll delete it.

File: main/plugins/org.talend.hadoop.distribution/src/main/java/org/talend/hadoop/distribution/HadoopDistributionPlugin.java (hunk 5 above)
Comment (cmeng-talend) [reply main/plugins/org.talend.hadoop.distribution/src/main/java/org/talend/hadoop/distribution/HadoopDistributionPlugin.java]: ok, I'll add it

**Pull request merged**